### PR TITLE
Extend at-time and at-elapsed

### PR DESCRIPTION
**Pull request by charleskawczynski**
This PR extends `Base.@time` / `Base.@elapsed` for CPU/CUDA devices.

### CODE REVIEW

**Review by simonbyrne — APPROVED** [review]

File: src/devices.jl [hunk 1]
```diff
@@ -109,3 +109,59 @@ macro threaded(device, expr)
...
+    return esc(quote
+        let
+            if $device isa ClimaComms.CUDADevice
+                CUDA.@time $(expr)
+            else
+                @assert $device isa ClimaComms.AbstractDevice
+                @time $(expr)
+            end
+        end
+    end)
```
Comment (simonbyrne) [inline src/devices.jl:138]:
```suggestion
    return quote
        let
            if $(esc(device)) isa ClimaComms.CUDADevice
                CUDA.@time $(esc(expr))
            else
                @assert $(esc(device)) isa ClimaComms.AbstractDevice
                Base.@time $(esc(expr))
            end
        end
    end
```

Comment (simonbyrne) [reply src/devices.jl:138]: I think should work

Comment (simonbyrne) [reply src/devices.jl:138]: not sure you need the `let`?

Comment (charleskawczynski) [reply src/devices.jl:138]: Looks good, I'm trying to write some hygiene tests now, I'll update the `@threaded` code first, then update this PR.

Comment (charleskawczynski) [reply src/devices.jl:138]: I'm running into an issue here (when updating the `@threaded` part):

`test/hygiene.jl`:
```julia
import ClimaComms: @threaded, CPUMultiThreaded, CPUSingleThreaded
dev = CPUSingleThreaded()
@threaded dev for i in 1:3
    1
end
```
error:
```julia
julia> using Revise; include("test/hygiene.jl")
ERROR: LoadError: ArgumentError: @threads requires a `for` loop expression
Stacktrace:
 [1] var"@threads"(__source__::LineNumberNode, __module__::Module, args::Vararg{Any})
   @ Base.Threads ./threadingconstructs.jl:226
 [2] include(fname::String)
   @ Base.MainInclude ./client.jl:476
 [3] top-level scope
   @ REPL[2]:1
in expression starting at /dev/CliMA/ClimaComms.jl/src/devices.jl:103
in expression starting at /dev/CliMA/ClimaComms.jl/test/hygiene.jl:3
```
where
```julia
macro threaded(device, expr)
    return quote
            if $(esc(device)) isa CPUMultiThreaded
                Threads.@threads $(esc(expr))
            else
                @assert $(esc(device)) isa CPUSingleThreaded
                $(esc(expr))
            end
        end
end
```

**Pull request merged**